src/TimeSpans.jl

```diff
@@ -275,4 +309,342 @@ function time_from_index(sample_rate, sample_range::AbstractUnitRange)
                     time_from_index(sample_rate, j))
 end
 
+#####
+##### set operations
+#####
+
+# represents a union of one or more time spans
+abstract type AbstractTimeSpanUnion <: AbstractVector{TimeSpan} end
+
+# Time span unions are read only; define a clear error message.
+function readOnlyError()
+    return error("This is a read only value. Call `collect` on the result to get " *
+                 "an editable copy or use one of the invariant preserving methods.")
+end
+
+# a `union` of a single time span
+# (an internal type that simplifies implementation of `mergesets`, below)
+struct TimeSpanSingleton <: AbstractTimeSpanUnion
+    data::TimeSpan
+end
+Base.@propagate_inbounds Base.getindex(x::TimeSpanSingleton, i...) = x.data
+Base.size(::TimeSpanSingleton) = (1,)
+Base.setindex(::TimeSpanSingleton, _) = readOnlyError()
+
+"""
+    TimeSpanUnion <: AbstractVector{TimeSpan}
+
+All set operations (e.g. `union`, `intersect`) over a `TimeSpan` or a
+`AbstractVector{TimeSpan}` return an object of type `TimeSpanUnion`.
+`TimeSpanUnion` contains some number of time spans and maintains
+the following invariants.
+
+1. Time spans do not overlap
+2. Time spans are sorted from earlier to later start times
+
+Using this type helps to ensure that sequences of multiple set operations do not
+have to repeatedly check and preserve these invariants.
+
+If you need to modify these sets yourself you can use one of the invariant
+preserving methods (`shrinkall!`, `translateall!`, etc...) or you can call
+`collect` and modify the resulting copy. Future calls to set operations
+will have to re-establish the invariants of such a copy.
+"""
+# a `union` of multiple time spans
+struct TimeSpanUnion{A} <: AbstractTimeSpanUnion
+    data::A
+    function TimeSpanUnion(x::AbstractVector{TimeSpan}, issorted=false,
+                           nooverlap=false)
```

Comment:
Okay, yeah that make sense. I think it is a little confusing to use then, becuase one of the arugments needs to be true and one false to indicate that all invariates hold; I've made them keyword arguments now to avoid confusion.